### PR TITLE
Td 7 UI

### PR DESCRIPTION
**Pull request by Apozzz**
8. Shop UI

a. Susipažinti su:

i. NodeJs,

ii. TypeScript

iii. Angular

iv. HTML/CSS

v. RxJS (Observable)

b. Parduotuvės programoje padaryti Parduotuvės UI

i. Prekių sąrašas

1. Rušiavimas

2. Filtravimas

ii. Prekės puslapis (info ir pardavimo mygtukas)
iv. Pakeisti CSS į SCSS
v. CSS iš npm (bootstrap or etc.)

### CODE REVIEW

File: ui/pom.xml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>com.twoday</groupId>
+        <artifactId>warehouse-management-system</artifactId>
+        <version>0.0.1-SNAPSHOT</version>
+    </parent>
+    <groupId>com.twoday.wms</groupId>
+	<artifactId>ui</artifactId>
+	<version>0.0.1-SNAPSHOT</version>
+	<name>frontend</name>
```
Comment (ArnasPetraitis) [inline ui/pom.xml]: which frontend? 😀 if only it was named `shop-ui` 🤔

File: ui/pom.xml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>com.github.eirslett</groupId>
+                <artifactId>frontend-maven-plugin</artifactId>
+                <version>1.13.4</version>
+                <executions>
+                    <execution>
+                        <id>install node and npm</id>
```
Comment (ArnasPetraitis) [inline ui/pom.xml]: nice find 👍

File: .gitignore [hunk 3]
```diff
@@ -29,7 +29,48 @@ build/
...
+# Visual Studio Code
+.vscode/*
+!.vscode/settings.json
+!.vscode/tasks.json
+!.vscode/launch.json
+!.vscode/extensions.json
+.history/*
+
+# Compiled output
+/dist
```
Comment (ArnasPetraitis) [inline .gitignore]: if started with `/` it looks from root, and your app is in subfolder, that's why `dist` and `.vscode` was not ignored

File: ui/src/main/frontend/src/app/components/product-filter/product-filter.component.less [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    overflow: hidden;
+}
+
+.mat-sidenav-container {
+    flex: 1;
+    top: 70px;
+    overflow: hidden;
+}
+
+mat-slider {
```
Comment (ArnasPetraitis) [inline ui/src/main/frontend/src/app/components/product-filter/product-filter.component.less]: `.` missing?

File: ui/src/main/frontend/src/app/components/product-filter/product-filter.component.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      this.maxQuantity = this.filterService.getMaxQuantity(this.productData);
+
+      this.priceRange = [this.minPrice, this.maxPrice];
+      this.quantityRange = [this.minQuantity, this.maxQuantity];
+    }
+  }
+
+  onFilterChange() {
+    this.filterChangedSubject.next({
+      searchName: this.searchName,
```
Comment (ArnasPetraitis) [inline ui/src/main/frontend/src/app/components/product-filter/product-filter.component.ts]: should be a type (class)

File: ui/src/main/frontend/src/app/components/product-list/product-list.component.html [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  <app-product-filter *ngIf="products?.length" [productData]="products"
+    (filterChanged)="onFilterChange($event)"></app-product-filter>
+  <div class="products-container">
+    <div class="products-list">
+      <h1>Products Results From Warehouse</h1>
+      <app-product-sort (sortChanged)="onSortChange($event)"></app-product-sort>
+      <div class="cards-container">
+        <mat-card *ngFor="let product of filteredProducts" class="product-card">
+          <mat-card-header>
+            <mat-card-title>ID: {{product.id}}</mat-card-title>
```
Comment (ArnasPetraitis) [inline ui/src/main/frontend/src/app/components/product-list/product-list.component.html]: very uncommon to show IDs for user

File: ui/src/main/frontend/src/app/components/product-sort/product-sort.component.html [hunk 7]
```diff
@@ -0,0 +1,12 @@
+<div class="sorting-container">
+    <mat-form-field appearance="fill">
+      <mat-label>Sort By</mat-label>
+      <mat-select [(value)]="sortOption" (selectionChange)="onSortChange()">
+        <mat-option value="id-asc">ID (Low to High)</mat-option>
```
Comment (ArnasPetraitis) [inline ui/src/main/frontend/src/app/components/product-sort/product-sort.component.html]: sort by name?

File: ui/src/main/frontend/src/app/components/product-sort/product-sort.component.less [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        font-size: 16px;
+        margin-right: 1em;  // spacing between label and dropdown
+        font-weight: 500;
+    }
+
+    select {
+        border: none;
+        padding: 0.5em 1em;
+        border-radius: 4px;
+        background: #f7f7f7;  // slightly different background for contrast
```
Comment (ArnasPetraitis) [inline ui/src/main/frontend/src/app/components/product-sort/product-sort.component.less]: if less supports variables, maybe extract all colors to common place to  have a consistent look

File: ui/src/main/frontend/src/app/services/product.service.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  }
+  
+  purchaseProduct(productId: number, quantity: number): Observable<any> {
+    const purchaseUrl = `${this.baseUrl}/${productId}/purchase?quantity=${quantity}`;
+    return this.httpClient.post(purchaseUrl, {});
+  }
+
+}
+
+export interface ProductDto {
```
Comment (ArnasPetraitis) [inline ui/src/main/frontend/src/app/services/product.service.ts]: maybe should be separated to own ts

File: ui/src/main/frontend/src/index.html [hunk 10]
```diff
@@ -0,0 +1,17 @@
+<!doctype html>
+<html lang="en">
+<head>
+  <meta charset="utf-8">
+  <title>YourAngularApp</title>
+  <base href="/">
+  <meta name="viewport" content="width=device-width, initial-scale=1">
+  <link rel="icon" type="image/x-icon" href="favicon.ico">
+  <link rel="preconnect" href="https://fonts.gstatic.com">
+  <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@300;400;500&display=swap" rel="stylesheet">
```
Comment (ArnasPetraitis) [inline ui/src/main/frontend/src/index.html]: external resources might cause unexpected problems. so usually resources are imported and used from npm

**Review by ArnasPetraitis — CHANGES_REQUESTED** [review]

File: ui/src/main/frontend/src/app/components/product-filter/product-filter.component.less (hunk 4 above)
Comment (Apozzz) [reply ui/src/main/frontend/src/app/components/product-filter/product-filter.component.less]: Cia viskas gerai. mat-slider yra elementas, ne klase

File: ui/src/main/frontend/src/app/components/product-sort/product-sort.component.html (hunk 7 above)
Comment (Apozzz) [reply ui/src/main/frontend/src/app/components/product-sort/product-sort.component.html]: Nelabai norejau, kazkaip niekad neteko naudoti ir atrodo nereikalinga(pagal pavadinima). Galiu nebent isimti pagal id.

File: ui/src/main/frontend/src/app/components/product-list/product-list.component.html [hunk 11]
```diff
@@ -0,0 +1,23 @@
+<table>
+  <caption>Products Results From Warehouse</caption>
+  <thead>
+    <tr>
+      <th>ID</th>
```
Comment (JustasVit) [inline ui/src/main/frontend/src/app/components/product-list/product-list.component.html]: Needs localization

Comment (JustasVit) [inline ui/src/main/frontend/src/app/components/product-list/product-list.component.html]: Do we really want to show ID?

**Review by JustasVit — CHANGES_REQUESTED** [review]

File: ui/src/main/frontend/src/app/components/product-list/product-list.component.html (hunk 11 above)
Comment (Apozzz) [reply ui/src/main/frontend/src/app/components/product-list/product-list.component.html]: Localization bus kitam PR :D

File: shop-ui/src/main/frontend/src/app/components/product-list/product-list.component.less [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+  flex-direction: column;
+  gap: @spacing-small;
+
+  .products-list {
+    display: flex;
+    flex-direction: column;
+    gap: @spacing-small;
+  }
+
+  h1 {
```
Comment (JustasVit) [inline shop-ui/src/main/frontend/src/app/components/product-list/product-list.component.less]: Don't style tags directly, use classes for that.

File: shop-ui/src/main/frontend/src/app/components/product-filter/product-filter.component.less [hunk 13]
```diff
@@ -19,27 +22,27 @@
...
 .filter-title {
     font-weight: 600;
     margin-bottom: 16px;
-    color: #333;
+    color: @color-gray;
 }
 
 .filter-input {
-    width: 40%;
+    width: @filter-input-width;
```
Comment (JustasVit) [inline shop-ui/src/main/frontend/src/app/components/product-filter/product-filter.component.less]: Create more generic variable for this.

File: shop-ui/src/main/frontend/src/app/shared/models/productDto.interface.ts [hunk 14]
```diff
@@ -0,0 +1,9 @@
+export interface ProductDto {
+    id: number;
+    name: string;
+    description: string;
```
Comment (JustasVit) [inline shop-ui/src/main/frontend/src/app/shared/models/productDto.interface.ts]: Shouldn't description be optional? In this case, if the description is missing, show different text in UI accordingly.

**Review by JustasVit — CHANGES_REQUESTED** [review]

File: shop-ui/src/main/frontend/src/app/components/product-sort/product-sort.component.less [hunk 15]
```diff
@@ -1,34 +1,36 @@
+@import "../../../styles/base/colors.less";
+
 .sorting-container {
     display: flex;
     align-items: center;
     justify-content: flex-start;
     padding: 1em;
-    background: #ffffff;
+    background: @color-white;
```
Comment (ArnasPetraitis) [inline shop-ui/src/main/frontend/src/app/components/product-sort/product-sort.component.less]: `@color-white` is just a bit better than `#ffffff`. it only alows nice transition to different shade. `@primary-backgroud-color` is more flexible

File: shop-ui/src/main/frontend/src/app/components/product-sort/product-sort.component.less [hunk 16]
```diff
@@ -1,34 +1,36 @@
...
     border-radius: 8px;
-    box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);  // subtle shadow
-    transition: all 0.3s ease;  // smooth transition for hover effects
+    box-shadow: 0 4px 6px @color-shadow-light;
```
Comment (ArnasPetraitis) [inline shop-ui/src/main/frontend/src/app/components/product-sort/product-sort.component.less]: `@color-shadow-light;` 👍

**Review by ArnasPetraitis — CHANGES_REQUESTED** [review]

**Comment by JustasVit** [comment]
Make filtering and sorting logic generic.

File: shop-ui/src/main/frontend/src/app/components/generic-filter/generic-filter.component.html [hunk 17]
```diff
@@ -0,0 +1,33 @@
+<mat-sidenav-container class="sidenav-container">
```
Comment (JustasVit) [inline shop-ui/src/main/frontend/src/app/components/generic-filter/generic-filter.component.html]: Component naming. Maybe just leave it as filter component?

File: shop-ui/src/main/frontend/src/app/components/generic-sort/generic-sort.component.html [hunk 18]
```diff
@@ -0,0 +1,10 @@
+<div class="sorting-container">
```
Comment (JustasVit) [inline shop-ui/src/main/frontend/src/app/components/generic-sort/generic-sort.component.html]: Component naming

File: shop-ui/src/main/frontend/src/app/components/generic-sort/generic-sort.component.ts [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+import { isEqual } from 'lodash';
+
+@Component({
+  selector: 'app-generic-sort',
+  templateUrl: './generic-sort.component.html',
+  styleUrls: ['./generic-sort.component.less']
+})
+export class GenericSortComponent<T> implements OnInit, OnChanges {
+
+  private sortChangedSubject = new Subject<T[]>();
```
Comment (JustasVit) [inline shop-ui/src/main/frontend/src/app/components/generic-sort/generic-sort.component.ts]: naming

File: shop-ui/src/main/frontend/src/app/components/product-list/product-list.component.ts [hunk 20]
```diff
@@ -14,53 +14,88 @@ import { ProductDto } from 'src/app/shared/models/product-dto.interface';
...
+  sortOptions: SortOption[] = [];
+  sortConfigs: SortConfig<ProductDto>[] = [];
+  filterFields: Array<keyof ProductDto> = [];
   private unsubscribe$ = new Subject<void>();
 
-  constructor(private productService: ProductService, private productFilteringService: ProductFilteringService, private productSortingService: ProductSortingService, private router: Router) { }
+  @ViewChild('sortComp', { static: false }) sortComponent!: GenericSortComponent<any>;
+
+  constructor(
+    //private productService: ProductService,
```
Comment (JustasVit) [inline shop-ui/src/main/frontend/src/app/components/product-list/product-list.component.ts]: remove

File: shop-ui/src/main/frontend/src/app/components/product-list/product-list.component.ts [hunk 21]
```diff
@@ -14,53 +14,88 @@ import { ProductDto } from 'src/app/shared/models/product-dto.interface';
...
-  private sortProductsByDefault(): void {
-    this.onSortChange('name-asc');
-  }
-
   ngOnDestroy(): void {
     this.unsubscribe$.next();
     this.unsubscribe$.complete();
   }
 
+  setFilterFields(): void {
```
Comment (JustasVit) [inline shop-ui/src/main/frontend/src/app/components/product-list/product-list.component.ts]: No need to have separate functions like setSortConfigs, setSortOptions etc. in your case.

File: shop-ui/src/main/frontend/src/app/components/purchase/purchase.component.html [hunk 22]
```diff
@@ -1,7 +1,7 @@
 <div class="main-container">
     <div class="form-container">
-        <form (ngSubmit)="executePurchase()">
-            <mat-form-field>
+        <form (ngSubmit)="executePurchase()" class="purchase-form">
```
Comment (JustasVit) [inline shop-ui/src/main/frontend/src/app/components/purchase/purchase.component.html]: purchase()

**Review by JustasVit — CHANGES_REQUESTED** [review]

**Review by JustasVit — APPROVED** [review]

**Pull request merged**